services/gitlab/gitlab-merge-requests.service.js

```diff
@@ -0,0 +1,359 @@
+      pattern: 'all-raw/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: {
+        labels: 'test,failure::new',
+        gitlab_url: 'https://gitlab.com',
+      },
+      staticPreview: {
+        label: 'all test,failure::new merge requests',
+        message: '12',
+        color: 'yellow',
+      },
+      documentation: documentation + labelDocumentation,
+    },
+    {
+      title: 'GitLab locked merge requests',
+      pattern: 'locked/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: { gitlab_url: 'https://gitlab.com' },
+      staticPreview: {
+        label: 'merge requests',
+        message: '0 locked',
+        color: 'yellow',
+      },
+      documentation,
+    },
+    {
+      title: 'GitLab locked merge requests by-label',
+      pattern: 'closed/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: {
+        labels: 'test,type::feature',
+        gitlab_url: 'https://gitlab.com',
+      },
+      staticPreview: {
+        label: 'test,failure::new merge requests',
+        message: '0 locked',
+        color: 'yellow',
+      },
+      documentation: documentation + labelDocumentation,
+    },
+    {
+      title: 'GitLab merged merge requests',
+      pattern: 'merged/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: { gitlab_url: 'https://gitlab.com' },
+      staticPreview: {
+        label: 'merge requests',
+        message: 'more than 10k merged',
+        color: 'brightgreen',
+      },
+      documentation,
+    },
+    {
+      title: 'GitLab merged merge requests by-label',
+      pattern: 'merged/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: {
+        labels: 'test,type::feature',
+        gitlab_url: 'https://gitlab.com',
+      },
+      staticPreview: {
+        label: 'test,failure::new merge requests',
+        message: '68 merged',
+        color: 'brightgreen',
+      },
+      documentation: documentation + labelDocumentation,
+    },
+  ]
+
+  static defaultBadgeData = { label: 'merge requests' }
+
+  static render({ variant, raw, labels, mergeRequestCount }) {
+    const state = variant
+    const isMultiLabel = labels && labels.includes(',')
+    const labelText = labels ? `${isMultiLabel ? `${labels}` : labels} ` : ''
+
+    let labelPrefix = ''
+    let messageSuffix = ''
+    if (raw !== undefined) {
+      labelPrefix = `${state} `
+    } else {
+      messageSuffix = state
+    }
+    const message = `${mergeRequestCount > 10000 ? 'more than ' : ''}${metric(
+      mergeRequestCount
+    )}${messageSuffix ? ' ' : ''}${messageSuffix}`
+    return {
+      label: `${labelPrefix}${labelText}merge requests`,
+      message,
+      color: mergeRequestCount > 0 ? 'brightgreen' : 'yellow',
+    }
+  }
```

Comment:
This function is fairly complex, largely because it contains a number of inline conditionals that leverage the ternary operator. I think it might help improve readability to break those up a bit, even if it does make things a tad more verbose.

Additionally, I don't think we should be applying custom colorization here. This badge strikes me as a purely informational data point and as such I feel like our `blue`/`informational` color would be most appropriate. I'm aware that the analog badges for GitHub utilize the yellow coloring, but as I think we may have discussed before I think those badges are colored that way for reasons that aren't necessarily applicable for GitLab.

Finally, given the breadth of values this function could produce, I think it would be best to do some unit testing of this functions to cover all those values and paths. Let me know if you're unfamiliar with how we do unit tests (as opposed to the service tests), and I'll be happy to share some additional info!

Reply:
Indeed, I've written a lot of unit tests for Go or Python, but never tried to write unit tests for JS functions, thanks for your help.

Reply:
Absolutely! I've found Go unit tests to be a bit odd compared to other languages, but I think you'll find the JS tests will feel fairly familiar to Python (albeit with some different conventions and apis).

For a pretty basic example I'd suggest taking a look at the unit tests for the codecov service:

https://github.com/badges/shields/blob/fa3839cdab3a7fc7b15c1cd35b4da175d4e28b07/services/codecov/codecov.spec.js#L1-L23

One difference with JS is that the actual "test" functions are typically implemented as callback functions that are passed as a call arg to the hooks/functions exported by the framework (as opposed to the free function defs, or test classes you'd define for tests in python), and inside that callback is where you'll see the typical arrange/act/assert paradigm.

For our unit tests we leverage a tool called [sazerac](https://github.com/sazeracjs/sazerac) which takes care of a lot of the boilerplate code for writing tests and provides a data-driven type of interface. With this we use a `describe` block (this just helps organize tests and makes it easier to run these subset of tests) and then we define a number of `test` calls for each respective function we want to test on the service class, almost always for a `render` and/or `transform` type of function.

The first arg to the `test` call should be the service class function, and the second is the callback that contains the test. The body of that callback will contain one or more `given(...)` calls, where the arg passed to `given` is the arg that will be passed to the service class function as part of the test. The chained `.expect()` call is where you'd specify the expected return value from calling that service class function with the arg(s) you specified for the associated `given` call.

I'm not sure if this will be applicable here, but in cases where passing multiple different inputs to the function under test should produce the same result, you can use the `forCases` function which just avoids the need to have multiple `.expect()` calls.

You can run these tests with `npm run test:core -- --grep=FooBar` where FooBar would be whatever value you name the suite in the `describe(...)` call, e.g. `npm run test:core -- --grep=Codecov` (note this is case sensitive, I forget if there's a case insensitive filter)

Reply:
awesome!